cdap-security/src/main/java/co/cask/cdap/security/store/DefaultSecureStoreService.java

```diff
@@ -95,68 +95,62 @@
    */
   @Override
-  public SecureStoreData get(SecureKeyId secureKeyId) throws Exception {
+  public final SecureStoreData getSecureData(String namespace, String name) throws Exception {
     Principal principal = authenticationContext.getPrincipal();
     final Predicate<EntityId> filter;
     filter = authorizationEnforcer.createFilter(principal);
-
+    SecureKeyId secureKeyId = new SecureKeyId(namespace, name);
     if (filter.apply(secureKeyId)) {
-      return secureStore.getSecureData(secureKeyId.getNamespace(), secureKeyId.getName());
+      return secureStore.getSecureData(namespace, name);
     }
     throw new UnauthorizedException(principal, Action.READ, secureKeyId);
   }
 
   /**
    * Puts the user provided data in the secure store, if the user has write access to the namespace. Grants the user
    * all access to the newly created entity.
-   * @param secureKeyId The Id for the key that needs to be stored.
-   * @param secureKeyCreateRequest The request containing the data to be stored in the secure store.
    * @throws BadRequestException If the request does not contain the value to be stored.
    * @throws UnauthorizedException If the user does not have write permissions on the namespace.
    * @throws NamespaceNotFoundException If the specified namespace does not exist.
    * @throws AlreadyExistsException If the key already exists in the namespace. Updating is not supported.
    * @throws IOException If there was a problem storing the key to underlying provider.
    */
   @Override
-  public synchronized void put(SecureKeyId secureKeyId,
-                               SecureKeyCreateRequest secureKeyCreateRequest) throws Exception {
+  public final synchronized void putSecureData(String namespace, String name, String value, String description,
+                                         Map<String, String> properties) throws Exception {
     Principal principal = authenticationContext.getPrincipal();
-    NamespaceId namespaceId = new NamespaceId(secureKeyId.getNamespace());
+    NamespaceId namespaceId = new NamespaceId(namespace);
     authorizationEnforcer.enforce(namespaceId, principal, Action.WRITE);
 
-    String description = secureKeyCreateRequest.getDescription();
-    String value = secureKeyCreateRequest.getData();
     if (Strings.isNullOrEmpty(value)) {
       throw new BadRequestException("The data field should not be empty. This is the data that will be stored " +
                                       "securely.");
     }
 
-    secureStoreManager.putSecureData(secureKeyId.getNamespace(), secureKeyId.getName(), value, description,
-                           secureKeyCreateRequest.getProperties());
-    authorizer.grant(secureKeyId, principal, ImmutableSet.of(Action.ALL));
+    secureStoreManager.putSecureData(namespace, name, value, description, properties);
+    privilegesManager.grant(new SecureKeyId(namespace, name), principal, ImmutableSet.of(Action.ALL));
```

Comment:
should grant first. See https://wiki.cask.co/display/CE/Authorization+-+CDAP+3.5#Authorization-CDAP3.5-AuthorizationPolicies. Also revoke all on the key as the first step